src/snooping.jl

```diff
@@ -1,94 +1,156 @@
 using Pkg, Serialization
 
-# Taken from SnoopCompile, modifying just `julia_cmd`
-function snoop_vanilla(flags, filename, commands, pwd)
-    println("Launching new julia process to run commands...")
-    # addprocs will run the unmodified version of julia, so we
-    # launch it as a command.
-    code_object = """
-            using Serialization
-            while !eof(stdin)
-                Core.eval(Main, deserialize(stdin))
-            end
-            """
+function get_dependencies(package_root)
+    project = joinpath(package_root, "Project.toml")
+    if !isfile(project)
+        error("Your package needs to have a Project.toml for static compilation. Please read here how to upgrade:")
+    end
+    project_deps = Pkg.TOML.parsefile(project)["deps"]
+    Symbol.(keys(project_deps))
+end
+
+"""
+Recursively get's all dependencies for `package` and makes sure they're installed!
+Note, that this will mutate your current Pkg environment!
+Returns all dependencies!
+"""
+function recursive_install_dependencies(package::Symbol, deps = Set{Symbol}(), visited = Set{Symbol}(); installed = Pkg.installed())
+  package in visited && return deps
+  if !haskey(installed, string(package))
+    @info "installing $package"
+    Pkg.add(string(package))
+  end
+  push!(visited, package)
+  M = @eval Module() ((import $package; $package))
+  project = joinpath(dirname(pathof(M)), "..", "Project.toml")
+  require = joinpath(dirname(pathof(M)), "..", "REQUIRE")
+  if isfile(project)
+    toml = Pkg.TOML.parsefile(project)
+    if haskey(toml, "deps")
+      push!(deps, Symbol.(keys(toml["deps"]))...)
+    end
+  elseif isfile(require)
+    for line in eachline(require)
+      isempty(line) && continue
+      any(x-> occursin(x, line), (
+        "windows", "osx", "unix", "linux", "julia", "#"
+      )) && continue
+      m = match(r"([a-zA-Z\-]+)", line)
+      if m == nothing
+        @warn("Can't match package: $line") # just means we got our regex wrong
+      else
+        push!(deps, Symbol(m[1]))
+      end
+    end
+  else
+    error("No Project.toml or REQUIRE found for package $package")
+  end
+  foreach(x-> recursive_install_dependencies(x, deps, visited; installed = installed), copy(deps))
+  deps
+end
+
+
+function snoop(package, snoopfile, outputfile; additional_packages = Symbol[])
+    command = """
+    using Pkg, $package
+    using $(join(additional_packages, ", "))
```

Comment:
Because I don't have a toml with all recursive deps. But I admit, a cleaner fix would be to generated a toml with all recursive deps and activate that! At this early stage, I was just trying to quickly get things running to see if it even works. If you have some time, it would be great if you can help with the Pkg part of this, since I'm sure your solutions will be much leaner ;) I'm still in a state, where I'm not a 100% sure how to nicely put together a full Pkg3 solution!